Ratpoison Test

TVsIan opened the pull request:
TEST ONLY - this currently replaces OpenBox with Ratpoison.

There is code in videomode.py to set up multi-frame displays and a reset script to return to ES afterwards. They're currently unused until I bring my previous changes back to the GC generator and set up Cemu multi-window display.

The features for Dolphin and Cemu are in place, and a new tag has been added to es_features.yml that will allow individual options to be restricted to certain arches, currently this only applied to the entire feature. s922x is the only arch that currently builds Dolphin but doesn't include GBA, but this is available for other systems that for example, may not support Vulkan.

My only current issue is with Mega Bezels, and that may be an unrelated problem on my end. Please test and see if there are any underlying issues with using Ratpoison over Openbox.

Darknior commented:
Woaw really cool options, it's a big improvement for cemu too :)
I really wish some one will found a solution for Mega Bezel, or turn it off when using this feature.
Thanks

nadenislamarre commented:
ifeq ($(BR2_PACKAGE_XORG7),y)
    DOLPHIN_EMU_CONF_OPTS += -DUSE_MGBA=ON
else
    DOLPHIN_EMU_CONF_OPTS += -DUSE_MGBA=OFF
endif

why ?

nadenislamarre commented:
LC_ALL=C ratpoison -c remhook deletewindow "execa ratpoison-reset"
LC_ALL=C ratpoison -c remhook newwindow next
LC_ALL=C ratpoison -c "fselect 1"


===>

export LC_ALL=C
ratpoison -c remhook deletewindow "execa ratpoison-reset"
ratpoison -c remhook newwindow next
ratpoison -c "fselect 1"

nadenislamarre commented:
i don't understand what this line does:
https://github.com/batocera-linux/batocera.linux/pull/7531/files#diff-cb08eb468c71bf5285ff338dbd7bea0e798966b5e2e8525738d6cb7db730c764R10

nadenislamarre commented:
i don't understand the option :
            choiceArch:
                "GBA (Internal)":             [x86, x86_64]

TVsIan commented:
> i don't understand the option : choiceArch: "GBA (Internal)": [x86, x86_64]

That, along with changes to batocera-es-system.py makes it so the GBA option will only be added to es_features.cfg on boards that support the GBA controllers.



> i don't understand what this line does: https://github.com/batocera-linux/batocera.linux/pull/7531/files#diff-cb08eb468c71bf5285ff338dbd7bea0e798966b5e2e8525738d6cb7db730c764R10

That's to allow for virtual desktops in ratpoison, it's from their Wiki (http://ratpoison.wxcvbn.org/cgi-bin/wiki.pl/Virtual_desktops_oneliner). I haven't tested it yet, that's next.



> LC_ALL=C ratpoison -c remhook deletewindow "execa ratpoison-reset" LC_ALL=C ratpoison -c remhook newwindow next LC_ALL=C ratpoison -c "fselect 1"
> 
> ===>
> 
> export LC_ALL=C ratpoison -c remhook deletewindow "execa ratpoison-reset" ratpoison -c remhook newwindow next ratpoison -c "fselect 1"

I can condense that down,. it was a last minute fix.



> ifeq ($(BR2_PACKAGE_XORG7),y) DOLPHIN_EMU_CONF_OPTS += -DUSE_MGBA=ON else DOLPHIN_EMU_CONF_OPTS += -DUSE_MGBA=OFF endif
> 
> why ?

Originally I had ratpoison and gba building on x86-based builds only. You had changed ratpoison to build on anything with xorg7 instead, and gba dependent on ratpoison. I think I had some trouble with the ratpoison dependency, then changed it to xorg since it ould cover the same boards. I should probably change that to x86 instead, since I don't think s922x (the only other board we build dolphin on) can handle it.

TVsIan commented:
I thought of a cleaner/better way to implement setting up the frames and
restoring the ES frame when done, so I’m working on that while setting up &
testing Dolphin.

On Sat, Nov 19, 2022 at 3:48 AM Nicolas Adenis-Lamarre <
***@***.***> wrote:

> i don't understand the option :
> choiceArch:
> "GBA (Internal)": [x86, x86_64]
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/7531#issuecomment-1320837128>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2VN37BQBB4OVN4PTSDWJCH6VANCNFSM6AAAAAASFAT3HE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

TVsIan commented:
Update:
* Replaced ratpoison-reset with batocera-ratpoison - switched to Python, new script is flexible enough to implement other functions if needed.
* Rewrote frame slicing function - now generates a string that can be converted to a frame set rather than running a series of slicing commands in ratpoison. The extra frame is a placeholder to prevent ES from stealing a visible frame.
* Eliminated the desktop switching code from ratpoisonrc, replaced with frameset saving/resetting in our scripts. Since we don't want to save the game's frameset, and only want to restore ES after exiting, it wasn't needed.
* GC-GBA functionality re-added, everything I had working on the previous commit should be working.

Note: batocera-screenshot works to capture the full screen with all windows in Ratpoison. We may want to replace the existing screenshot hotkey function with that in the .keys file, it looks like it's using Dolphin's internal one right now, which won't capture GBA screens.

I'm currently doing a partial rebuild just to make sure the set of scripts I just pushed are the 100% working set, as I was live editing my setup (and entering in commands manually to fix things), but these should match what was there on the final overlay save and test. Once that's done, I'll hit cemu.

nadenislamarre commented:
arg, sorry, i have not a lot of time these days.
i must review and fix things i dont like.

=> in launch.py : this is forbidden. names of emulors/system must be only in generators.
if systemName == 'gamecube' and 'auto_ereader' in system.config and system.config['auto_ereader'] != "none":

=> os.symlink(f"/userdata
symlinks are forbidden in /userdata

=> choiceArch remains a mystery, a probably not correct. what does it mean for the end user ?

=>  Evmapy.start('gamecube-ereader' is probably incorrect. what you want is that evmappy can be differernt for a same emulator depending on the  options set ?

TVsIan commented:
For launch.py & evmapy: yes, if the e-reader option is enabled, we want to
have P2’s analog stick control the mouse. Loading e-reader cards requires
mouse input to use the menus.

For the symlink: Dolphin’s GBA save file names do not match the GBA
emulators - it uses .sav instead of .srm, and attaches the slot number at
the end of the file. In order to use the GBA saves without users needing to
manually copy back and forth, it creates the link. I believe I can route
that to a temporary folder like /var/run/gba or something instead, since
new GBA saves shouldn’t be created.

For choiceArch: this prevents the GBA option from appearing on platforms
that do not support GBA. The other options would be to have it there but
not working, or create a parallel set of options and duplicate the
controller selection code in the generator.

On Wed, Nov 23, 2022 at 2:21 AM Nicolas Adenis-Lamarre <
***@***.***> wrote:

> arg, sorry, i have not a lot of time these days.
> i must review and fix things i dont like.
>
> => in launch.py : this is forbidden. names of emulors/system must be only
> in generators.
> if systemName == 'gamecube' and 'auto_ereader' in system.config and
> system.config['auto_ereader'] != "none":
>
> => os.symlink(f"/userdata
> symlinks are forbidden in /userdata
>
> => choiceArch remains a mystery, a probably not correct. what does it mean
> for the end user ?
>
> => Evmapy.start('gamecube-ereader' is probably incorrect. what you want is
> that evmappy can be differernt for a same emulator depending on the options
> set ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/batocera-linux/batocera.linux/pull/7531#issuecomment-1324644567>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE5KD2VZHBL5WM3OLG53EC3WJXAWRANCNFSM6AAAAAASFAT3HE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

TVsIan commented:
Would it be acceptable to pass the system config to evmapy, and change the
system there if the option is on? When we have multiplayer GBA, we can
implement the same key mapping, and have an “ereader.keys” instead of
“gamecube-ereader.keys.”

Related, I think standalone mGBA might work for multiplayer, VBA-M requires
menus to activate the link. But I will look into that once these functions
are settled.

On Wed, Nov 23, 2022 at 6:10 AM Ian Murray ***@***.***> wrote:

> For launch.py & evmapy: yes, if the e-reader option is enabled, we want to
> have P2’s analog stick control the mouse. Loading e-reader cards requires
> mouse input to use the menus.
>
> For the symlink: Dolphin’s GBA save file names do not match the GBA
> emulators - it uses .sav instead of .srm, and attaches the slot number at
> the end of the file. In order to use the GBA saves without users needing to
> manually copy back and forth, it creates the link. I believe I can route
> that to a temporary folder like /var/run/gba or something instead, since
> new GBA saves shouldn’t be created.
>
> For choiceArch: this prevents the GBA option from appearing on platforms
> that do not support GBA. The other options would be to have it there but
> not working, or create a parallel set of options and duplicate the
> controller selection code in the generator.
>
> On Wed, Nov 23, 2022 at 2:21 AM Nicolas Adenis-Lamarre <
> ***@***.***> wrote:
>
>> arg, sorry, i have not a lot of time these days.
>> i must review and fix things i dont like.
>>
>> => in launch.py : this is forbidden. names of emulors/system must be only
>> in generators.
>> if systemName == 'gamecube' and 'auto_ereader' in system.config and
>> system.config['auto_ereader'] != "none":
>>
>> => os.symlink(f"/userdata
>> symlinks are forbidden in /userdata
>>
>> => choiceArch remains a mystery, a probably not correct. what does it
>> mean for the end user ?
>>
>> => Evmapy.start('gamecube-ereader' is probably incorrect. what you want
>> is that evmappy can be differernt for a same emulator depending on the
>> options set ?
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/batocera-linux/batocera.linux/pull/7531#issuecomment-1324644567>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AE5KD2VZHBL5WM3OLG53EC3WJXAWRANCNFSM6AAAAAASFAT3HE>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

TVsIan commented:
Not currently at home to test, but this last change should take care of the GBA save path issue as long as Dolphin cooperates with the save path option. It's moved to one variable so it can be changed if needed.

nadenislamarre commented:
why only with ereader ?


>>For launch.py & evmapy: yes, if the e-reader option is enabled, we want to
>>have P2’s analog stick control the mouse. Loading e-reader cards requires
>>>mouse input to use the menus.
we could enable this all the time, no ?

nadenislamarre commented:
>>For choiceArch: this prevents the GBA option from appearing on platforms
>>that do not support GBA. The other options would be to have it there but
>>not working, or create a parallel set of options and duplicate the
>>controller selection code in the generator.
which platform doesn't support it ? i don't understand this option. can you give me an example of issue it causes please ?

nadenislamarre commented:
there are several points not ok. for me, we should merge step by step. first migrate to ratpoison.

TVsIan commented:
> why only with ereader ?
> 
> > > For launch.py & evmapy: yes, if the e-reader option is enabled, we want to
> > > have P2’s analog stick control the mouse. Loading e-reader cards requires
> > > > mouse input to use the menus.
> > > > we could enable this all the time, no ?

For the Gamecube, normally Player 2 will need both analog sticks for gameplay. When that controller is set to a GBA, the stick is available to use as a mouse since it only uses the D-Pad.

> > > For choiceArch: this prevents the GBA option from appearing on platforms
> > > that do not support GBA. The other options would be to have it there but
> > > not working, or create a parallel set of options and duplicate the
> > > controller selection code in the generator.
> > > which platform doesn't support it ? i don't understand this option. can you give me an example of issue it causes please ?

I currently have the GBA functionality for Dolphin set to only build on x86-based boards, due to the power needed to emulate the Gamecube plus 1-4 GBAs. For other boards, if you select GBA as the controller, you would wind up with no input for that slot. That addition plus changes to the script makes it so the GBA option doesn't show up on other boards, but all the other choices do.

> there are several points not ok. for me, we should merge step by step. first migrate to ratpoison.

We can do that, I would have to keep some changes to the Cemu generator and .keys file so that gamepad view works. If I just remove the new options for es_features.yml, that should go around all of those things.

TVsIan commented:
This last batch of changes should address the issues. I can still remove most of it, but this gives a test to the frame layouts.

* Removed the evmapy override. Instead it looks for that feature, and adds a P2 Mouse profile to the loaded list. This would also be used for multiplayer GBA for the same purpose - the P2 controller will control both the GBA with the e-reader attached and the menus for loading the card data.
* Changed the GBA save symlinking to a temporary folder on the system drive.
* Removed the per-system choice option from es_systems, and set GBA to build on all boards that build Dolphin.

nadenislamarre commented:
if that's ok for you, we can merge.

TVsIan commented:
That should be fine.
